### PR TITLE
Type/display code on mobile (use tabs)

### PR DESCRIPTION
**Pull request by gabriel**
See https://keybase.atlassian.net/browse/DESKTOP-2101

2016-11-15 at 3 05 pm

2016-11-15 at 3 09 pm

2016-11-15 at 3 28 pm

2016-11-15 at 3 31 pm

2016-11-15 at 3 31 pm

### CODE REVIEW

**Comment by gabriel** [comment]
@keybase/react-hackers

**Comment by cjb** [comment]
@gabriel oh, the tabs are only supposed to be shown when it's phone-to-phone provisioning, and not computer-to-phone or phone-to-computer, sorry I didn't tell you that

File: shared/common-adapters/tab-bar.desktop.js [hunk 1]
```diff
@@ -90,12 +90,12 @@ class TabBarButton extends Component<void, TabBarButtonProps, void> {
...
 
-  _renderIcon (color: string, badgeNumber: number) {
-    if (this.props.source.type !== 'icon') return // needed to make flow happy
+  _renderDefault (color: string, badgeNumber: number) {
+    if (this.props.source && this.props.source.type !== 'icon') return // needed to make flow happy
     const backgroundColor = this.props.selected ? globalColors.darkBlue4 : globalColors.midnightBlue
     return (
       <Box style={{...stylesTabBarButtonIcon, backgroundColor, ...this.props.style}}>
-        <Icon type={this.props.source.icon} style={{...stylesIcon, color, ...this.props.styleIcon}} />
+        {this.props.source && <Icon type={this.props.source.icon} style={{...stylesIcon, color, ...this.props.styleIcon}} />}
```
Comment (MarcoPolo) [inline shared/common-adapters/tab-bar.desktop.js]: You need this for flow even with the above check?

File: shared/common-adapters/tab-bar.desktop.js [hunk 2]
```diff
@@ -90,12 +90,12 @@ class TabBarButton extends Component<void, TabBarButtonProps, void> {
     )
   }
 
-  _renderIcon (color: string, badgeNumber: number) {
-    if (this.props.source.type !== 'icon') return // needed to make flow happy
+  _renderDefault (color: string, badgeNumber: number) {
```
Comment (MarcoPolo) [inline shared/common-adapters/tab-bar.desktop.js]: I don't understand the point of this change. This function renders the icon variant. It just happens to be the default, but it shouldn't be defined as being the default.

File: shared/common-adapters/tab-bar.js.flow [hunk 3]
```diff
@@ -31,9 +30,11 @@ export type ItemProps = Exact<{
 
 export class TabBarItem extends Component<void, ItemProps, void> { }
 
+export type TabBarButtonSource = {type: 'icon', icon: IconType} | {type: 'avatar', avatar: Avatar} | {type: 'nav', icon: IconType}
+
 export type TabBarButtonProps = Exact<{
-  selected: boolean,
-  source: {type: 'icon', icon: IconType} | {type: 'avatar', avatar: Avatar} | {type: 'nav', icon: IconType},
+  selected?: boolean,
+  source?: TabBarButtonSource,
```
Comment (MarcoPolo) [inline shared/common-adapters/tab-bar.js.flow]: I don't think it makes sense to make this optional. Maybe you should add another variant to what you are trying to do. Which seems like just a label?

File: shared/login/register/passphrase/index.render.native.js [hunk 4]
```diff
@@ -58,7 +58,6 @@ const stylesCard = {
 }
 const usernameStyle = {
   textAlign: 'center',
-  paddingBottom: globalMargins.small,
```
Comment (gabriel) [inline shared/login/register/passphrase/index.render.native.js]: This looked problematic with keyboard up on IPhone SE

File: shared/common-adapters/tab-bar.desktop.js (hunk 1 above)
Comment (gabriel) [inline shared/common-adapters/tab-bar.desktop.js]: Yeah I forgot that SimpleTabBarButton existed by default. Removed.

File: shared/common-adapters/tab-bar.js.flow (hunk 3 above)
Comment (gabriel) [inline shared/common-adapters/tab-bar.js.flow]: Yeah I forgot that SimpleTabBarButton existed by default. Removed.

**Comment by gabriel** [comment]
ok, ready for review

**Comment by keybase-ci-visdiff** [comment]
The commits e6abc06c259e...c2b8a31dcbfa introduce visual changes on linux. <details><summary>:mag_right: 8 new, 1 removed</summary>

 * Added: **Code_Page-Show_Text_Mobile**  
   Code_Page-Show_Text_Mobile rendered
 * Added: **Code_Page-Show_Text**  
   Code_Page-Show_Text rendered
 * Added: **Code_Page-Show_Code_Mobile**  
   Code_Page-Show_Code_Mobile rendered
 * Added: **Code_Page-Show_Code**  
   Code_Page-Show_Code rendered
 * Added: **Code_Page-Scan_Code_Mobile**  
   Code_Page-Scan_Code_Mobile rendered
 * Added: **Code_Page-Scan_Code**  
   Code_Page-Scan_Code rendered
 * Added: **Code_Page-Enter_Text_Mobile**  
   Code_Page-Enter_Text_Mobile rendered
 * Added: **Code_Page-Enter_Text** [(view)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1479494139851-e6abc06c259e-c2b8a31dcbfa-linux/Code_Page-Enter_Text-c2b8a31dcbfa.png)
 * Removed: **Code_Page-Normal** [(view original)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1479494139851-e6abc06c259e-c2b8a31dcbfa-linux/Code_Page-Normal-e6abc06c259e.png)
</summary>

File: shared/login/register/code-page/index.render.native.js [hunk 5]
```diff
@@ -8,222 +8,223 @@ import Container from '../../forms/container'
...
-      codePageModeShowCode: this.renderIntroShowQR(),
-      codePageModeShowText: this.renderIntroShowTextCode(),
-      codePageModeEnterText: this.renderIntroEnterTextCode(),
-    }
-
-    return headerTextMap[this.props.mode] || <Text type='BodySmall'>Good luck</Text>
-  }
-
-  renderShowText () {
+  renderShowCodeForDesktop () {
```
Comment (MarcoPolo) [inline shared/login/register/code-page/index.render.native.js]: When does this get called?

File: shared/login/register/code-page/index.render.native.js [hunk 6]
```diff
@@ -8,222 +8,223 @@ import Container from '../../forms/container'
...
+            {this.renderEnterText()}
+          </TabBarItem>
+        </TabBar>
+        {this.renderSwitchButton(codePageModeScanCode, 'icon-phone-qr-code-48', 'Scan QR code instead')}
       </Container>
     )
   }
+
+  render () {
+    let isMobile = this.props.otherDeviceRole === codePageDeviceRoleExistingPhone
```
Comment (MarcoPolo) [inline shared/login/register/code-page/index.render.native.js]: You are guaranteed to be in mobile since this is a `.native.js` file

Comment (gabriel) [inline shared/login/register/code-page/index.render.native.js]: It's whether the other device being used for provisioning is mobile. (Whether to show the tabbed views or not.)

File: shared/login/register/code-page/index.render.native.js (hunk 5 above)
Comment (gabriel) [inline shared/login/register/code-page/index.render.native.js]: I'm not sure... maybe you could show QR code to desktop via it's camera... Like put your phone up to the macbook camera.

I don't think we do that yet but, for completeness, I included it.

File: shared/login/register/code-page/index.render.native.js (hunk 6 above)
Comment (gabriel) [inline shared/login/register/code-page/index.render.native.js]: I changed var name to `otherDeviceIsMobile` to make that clearer.

**Comment by keybase-ci-visdiff** [comment]
The commits 4022693e454e...07342967ddb3 introduce visual changes on linux. <details><summary>:mag_right: 8 new, 1 removed</summary>

 * Added: **Code_Page-Show_Text_Mobile**  
   Code_Page-Show_Text_Mobile rendered
 * Added: **Code_Page-Show_Text**  
   Code_Page-Show_Text rendered
 * Added: **Code_Page-Show_Code_Mobile**  
   Code_Page-Show_Code_Mobile rendered
 * Added: **Code_Page-Show_Code**  
   Code_Page-Show_Code rendered
 * Added: **Code_Page-Scan_Code_Mobile**  
   Code_Page-Scan_Code_Mobile rendered
 * Added: **Code_Page-Scan_Code**  
   Code_Page-Scan_Code rendered
 * Added: **Code_Page-Enter_Text_Mobile**  
   Code_Page-Enter_Text_Mobile rendered
 * Added: **Code_Page-Enter_Text** [(view)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1479755064283-4022693e454e-07342967ddb3-linux/Code_Page-Enter_Text-07342967ddb3.png)
 * Removed: **Code_Page-Normal** [(view original)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1479755064283-4022693e454e-07342967ddb3-linux/Code_Page-Normal-4022693e454e.png)
</summary>

**Comment by keybase-ci-visdiff** [comment]
The commits 4d5f1b5c1466...f226ad8eb798 introduce visual changes on linux. <details><summary>:mag_right: 8 new, 1 removed</summary>

 * Added: **Code_Page-Show_Text_Mobile**  
   Code_Page-Show_Text_Mobile rendered
 * Added: **Code_Page-Show_Text**  
   Code_Page-Show_Text rendered
 * Added: **Code_Page-Show_Code_Mobile**  
   Code_Page-Show_Code_Mobile rendered
 * Added: **Code_Page-Show_Code**  
   Code_Page-Show_Code rendered
 * Added: **Code_Page-Scan_Code_Mobile**  
   Code_Page-Scan_Code_Mobile rendered
 * Added: **Code_Page-Scan_Code**  
   Code_Page-Scan_Code rendered
 * Added: **Code_Page-Enter_Text_Mobile**  
   Code_Page-Enter_Text_Mobile rendered
 * Added: **Code_Page-Enter_Text** [(view)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1480445101229-4d5f1b5c1466-f226ad8eb798-linux/Code_Page-Enter_Text-f226ad8eb798.png)
 * Removed: **Code_Page-Normal** [(view original)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1480445101229-4d5f1b5c1466-f226ad8eb798-linux/Code_Page-Normal-4d5f1b5c1466.png)
</summary>

**Pull request merged**